database2/ffldb/ldb/transaction.go

```diff
@@ -0,0 +1,94 @@
+package ldb
+
+import (
+	"errors"
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// LevelDBTransaction is a thin wrapper around native leveldb
+// batches and snapshots. It supports both get and put.
+// Note: transactions provide data consistency over the state of
+// the database as it was when the transaction started. There is
+// NO guarantee that if one puts data into the transaction then
+// it will be available to get within the same transaction.
+type LevelDBTransaction struct {
+	ldb      *leveldb.DB
+	snapshot *leveldb.Snapshot
+	batch    *leveldb.Batch
+
+	isClosed bool
+}
+
+// Begin begins a new transaction.
+func (db *LevelDB) Begin() (*LevelDBTransaction, error) {
+	snapshot, err := db.ldb.GetSnapshot()
+	if err != nil {
+		return nil, err
+	}
+	batch := new(leveldb.Batch)
+
+	transaction := &LevelDBTransaction{
+		ldb:      db.ldb,
+		snapshot: snapshot,
+		batch:    batch,
+
+		isClosed: false,
+	}
+	return transaction, nil
+}
+
+// Commit commits whatever changes were made to the database
+// within this transaction.
+func (tx *LevelDBTransaction) Commit() error {
+	if tx.isClosed {
+		return errors.New("cannot commit a closed transaction")
+	}
+
+	tx.isClosed = true
+	tx.snapshot.Release()
```

Comment:
Can you explain (or provide a link) how does the relationship between snapshot and batch works